### PR TITLE
fetchdata and fetchmulti

### PR DESCRIPTION
**Pull request by zimbatm**
###### Motivation for this change

Moving the fetch data into pure NIXON code means we can more easily write update scripts. The updating part is still missing though.

The second motivation is that it's currently painful to update sha256s with multi-arch sources (binary packages). With `fetchmulti` it's possible to do `nix-prefetch-url . -A mypackage.src.all` and get all the hashes out.
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

/cc @garbas @kamilchm @domenkozar

### CODE REVIEW

**Comment by garbas** [comment]
:+1:

i imagine there are many other situations where ``fetchmulti`` could be used. is there a way to grep for those cases?

also do we need to document this? are other fetch* functions documented?

**Comment by edolstra** [comment]
The name `fetchmulti` implies that it fetches multiple files. But if I understand it correctly, it fetches just one file from a selection. So it should be something like `fetchOneOf`.

File: pkgs/build-support/fetchdata/default.nix [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+#      sha256 = "xxx";
+#    }
+#
+# Which means that the attributes can now be imported
+#
+#    fetchdata (import ./source.nix);
+#
+# That's the whole point of this function.
+#
+{ fetcher ? "fetchurl", ... }@attrs:
```
Comment (edolstra) [inline pkgs/build-support/fetchdata/default.nix]: Passing functions as strings seems really ugly to me...

Also, there is nothing in the `fetchdata` function specific to fetching. Rather it's a generic "call a function specified by the 'type' argument". But such a function doesn't sound like a good idea to me. Why write `f { type = "g"; args... }` when I can write `g args` directly?

File: pkgs/applications/audio/renoise/default.nix [hunk 2]
```diff
@@ -1,36 +1,36 @@
...
+    "i686-linux-demo" = {
+      url = "http://files.renoise.com/demo/Renoise_3_0_1_Demo_x86.tar.bz2";
+      sha256 = "0dgqvib4xh2yhgh2wajj11wsb6xiiwgfkhyz32g8vnyaij5q8f58";
+    };
+    "i686-linux" = {
+      fetcher = "requireFile";
+      url = "http://backstage.renoise.com/frontend/app/index.html#/login";
+      name = "rns_3_0_1_reg_x86.tar.gz";
+      sha256 = "1swax2jz0gswdpzz8alwjfd8rhigc2yfspj7p8wvdvylqrf7n8q7";
+    };
```
Comment (grahamc) [inline pkgs/applications/audio/renoise/default.nix]: Whooaa! Cool!!

**Comment by zimbatm** [comment]
Thanks for the reviews! I realize my vision isn't complete here and it's more of a 20% review.

The problem I am trying to solve is that it takes a lot of time to update packages. Most of the time it's enough to bump the version and update the sha256 but right now it's hard to automate that because the data is mixed with the mkDerivation bits. First bump the version, then run `nix-prefetch-url` and copy-and-paste the sha256 back into the file. My thinking is that if that data would be externalized then it would make things easier to automate.

Right now all I am doing in this PR is to move all the data into a pure nix (NIXON) declaration (which explain why fetcher is also a string) with the idea that in a further PRs I will be able to move that out to a separate file. I also converted a couple of files to `fetchOneOf` to demonstrate how it would work but there are a lot left to do.

I want us to get into a situation where the source data can live in a separate file (eg: `fetchdata import ./source.nix`) and have a tool that's able to update that external file either automatically by finding the list of upstream releases or manually with the user giving the version that he wants. If we have that we can also notify the maintainers of new releases and get a better sense of which parts of nixpkgs need more work.

If we can agree on the `fetcher` as a key and that interface in general I am happy to just propose the `fetchOneOf` fetcher in this PR and wait for the more complete picture to form before introducing the `fetchdata` function into nixpkgs.

What do you think?

**Comment by zimbatm** [comment]
@garbas for now my approach to finding multi-arch binaries is to grep for `x86_64.*\.tar` but it's not perfect

**Comment by zimbatm** [comment]
Moving the discussion to #19582. I will come back with a more complete solution with my updating script idea.

File: pkgs/build-support/fetchdata/default.nix (hunk 1 above)
Comment (vcunat) [inline pkgs/build-support/fetchdata/default.nix]: Yeah, not very nice. Clearly the goal is to be able to write in "nixon", i.e. without function calls.

Comment (zimbatm) [inline pkgs/build-support/fetchdata/default.nix]: Would it be more compelling if I replaced it with a `type` attribute such that `type = "git"` would map to the `fetchgit` fetcher and so on?

**Pull request closed without merging**